File: buy_maximum_stocks.cpp
//{ Driver Code Starts
/* Driver program to test above function */

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
 public:
  int buyMaximumProducts(int n, int k, int price[]) {
    // Write your code here
    vector<pair<int, int>> v(n);
    for (int i = 0; i < n; ++i) {
      v[i] = make_pair(price[i], i + 1);
    }
    sort(v.begin(), v.end());
    int ans = 0;
    for (int i = 0; i < n; ++i) {
      if (k >= v[i].first * v[i].second) {
        ans += v[i].second;
        k -= v[i].first * v[i].second;
      } else {
        ans += k / v[i].first;
        break;
      }
    }
    return ans;
  }
};

//{ Driver Code Starts.
int main() {
  int t;
  cin >> t;
  while (t--) {
    int n, k;
    cin >> n >> k;
    int price[n];
    for (int i = 0; i < n; i++) {
      cin >> price[i];
    }
    Solution ob;
    int ans = ob.buyMaximumProducts(n, k, price);
    cout << ans << endl;
  }
  return 0;
}

// } Driver Code Ends

File: greedy/buy_maximum_stocks.py
from typing import List


class Solution:
    def buyMaximumProducts(self, n: int, k: int, price: List[int]) -> int:
        # code here
        stock = []
        for i, p in enumerate(price):
            stock.append((p, i + 1))

        stock.sort()
        ans = 0
        for s in stock:
            if k >= s[1] * s[0]:
                ans += s[1]
                k -= s[1] * s[0]
            else:
                ans += k // s[0]
                break

        return ans


# {
# Driver Code Starts


class IntArray:
    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n, k = map(int, input().strip().split())

        price = IntArray().Input(n)

        obj = Solution()
        res = obj.buyMaximumProducts(n, k, price)

        print(res)


# } Driver Code Ends
